Add support for Google Takeout Location History

postmaxin opened the pull request:
This adds support for parsing the location history provided by Google Takeout. This makes it possible to convert a month's, year's, or your entire location history at once instead of having to export one day's history at a time on Google Maps.

There is documentation in the `xmldoc` folder with some examples.

It works great, but there are one, possibly two, other additions I'd like to make in following diffs:

1. Add the ability to select date ranges. For example, I went on a road trip that started in the last week of February, but right now I can only select tracks at the resolution of a whole month.

2. Add support for the higher-resolution "roadSegment" blocks in newer "activitySegment" blocks. This will have to be optional as you have to query the Google Maps API to get lat/long, which costs money. However, if you're willing to pay for it, you can get great detail out of it. They look like this:

```
       "roadSegment": [{
          "placeId": "ChIJBdXmuuWaK4gRoiClERhZcy8",
          "duration": "76s"
        }, {
          "placeId": "ChIJ4ekNlOWaK4gRxu-moMRFo60",
          "duration": "58s"
        }, {
          "placeId": "ChIJ10qEheWaK4gRQ1T24wEuuN4",
          "duration": "58s"
        }, {
          "placeId": "ChIJe7NWfu-aK4gRFHBB6P0PMe4",
          "duration": "58s"
[...]
```

postmaxin committed:
Add support for Google Takeout Location History

This adds support for parsing the location history provided by Google Takeout.
This makes it possible to convert a month's, year's, or your entire location
history at once instead of having to export one day's history at a time on
Google Maps.

There is documentation in the `xmldoc` folder with some examples.

It works great, but there are one, possibly two, other additions I'd like to
make in following diffs:

1. Add the ability to select date ranges. For example, I went on a road trip
   that started in the last week of February, but right now I can only select
	 tracks at the resolution of a whole month.

2. Add support for the higher-resolution "roadSegment" blocks in newer
   "activitySegment" blocks. This will have to be optional as you have to
	 query the Google Maps API to get lat/long, which costs money. However,
	 if you're willing to pay for it, you can get great detail out of it.
	 They look like this:

```
       "roadSegment": [{
          "placeId": "ChIJBdXmuuWaK4gRoiClERhZcy8",
          "duration": "76s"
        }, {
          "placeId": "ChIJ4ekNlOWaK4gRxu-moMRFo60",
          "duration": "58s"
        }, {
          "placeId": "ChIJ10qEheWaK4gRQ1T24wEuuN4",
          "duration": "58s"
        }, {
          "placeId": "ChIJe7NWfu-aK4gRFHBB6P0PMe4",
          "duration": "58s"
[...]
```

postmaxin committed:
rename variable

postmaxin committed:
this TODO is done

postmaxin committed:
now builds on focal, fix documentation

postmaxin committed:
patch -p2 < ~/testo.log

postmaxin committed:
use `&&` instead of `and` for old compilers

postmaxin committed:
don't leak objects when we skip events at exact lat/lon 0/0

postmaxin committed:
use a reference (even though the compiler would have optimized that away)

postmaxin committed:
rename to googletakeout

postmaxin committed:
fix docs

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+{
+  Waypoint* waypoint = new Waypoint();
+  waypoint->latitude = (double)(lat_e7 / 1e7);
+  waypoint->longitude = (double)(lon_e7 / 1e7);
```
these casts are redundant, clang-tidy flags:
 warning: redundant cast to the same type [google-readability-casting]

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+  const QJsonValue timelineObjectsIn = root.value(TIMELINE_OBJECTS);
+  if (timelineObjectsIn.isNull()) {
+    _fatal(
+      QString(ifd->fileName()) + " is missing required \"" +
```
ifd->filename() is a QString already.

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+    if (val.isObject()) {
+      timeline.append(val.toObject());
+    } else {
+      _fatal(QString(ifd->fileName()) + " has non-object in timelineObjects");
```
ifd->filename() is a QString

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+   *  For now, we try to use (2), and if (2) is not available, we fall
+   *  back on (1).
+   */
+  if (false && activitySegment.contains(SIMPLE_PATH)) {
```
This code is dead

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+  return n_points;
+}
+
+GoogleTakeoutInputStream::GoogleTakeoutInputStream() {}
```
use = default (in the .h file).
warning: use '= default' to define a trivial default constructor [hicpp-use-equals-default,modernize-use-equals-default]

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+// we use "latE7"/"lngE7" here instead of "latitudeE7"/"longitudeE7".
+// +10 points for brevity, but -100 points for inconsistency.
+#define LATE7 "latE7"
+#define LONE7 "lngE7"
```
can we make these private static class constants?
 used to declare a constant; consider using a 'constexpr' constant [cppcoreguidelines-macro-usage]

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+#define LATE7 "latE7"
+#define LONE7 "lngE7"
+
+const QList<QString> month_names{
```
can we make this a private constant as well so it doesn't pollute the global namespace?

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+      title[i] = title[i].toLower();
+    }
+  }
+}
```
can these be private member functions?
warning: declaration uses identifier '_warning', which is reserved in the global namespace [bugprone-reserved-identifier,cert-dcl37-c,cert-dcl51-cpp]
warning: declaration uses identifier '_fatal', which is reserved in the global namespace [bugprone-reserved-identifier,cert-dcl37-c,cert-dcl51-cpp]

postmaxin replied:
`_fatal` and `_warning` are used by both classes. i'll rename them.

tsteven4 replied:
or make GoogleTakeoutInputStream a nested class of GoogleTakeoutFormat

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+static void title_case(QString& title)
+{
+  bool new_word = true;
+  for (int i = 0; i < title.size(); ++ i) {
```
warning: use range-based for loop instead [modernize-loop-convert]
  for (int i = 0; i < title.size(); ++ i) {
  ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      (auto & i : title)

tsteven4 commented on CMakeLists.txt:
```diff
@@ -87,6 +87,7 @@ set(ALL_FMTS ${MINIMAL_FMTS}
   geocache.cc
   geojson.cc
   globalsat_sport.cc
+  googletakeout.cc
```
please add googletakeout.h to the HEADERS as well.

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+
+/* create a waypoint from late7/lone7 and optional metadata */
+Waypoint*
+GoogleTakeoutFormat::_waypoint(
```
warning: method '_waypoint' can be made static [readability-convert-member-functions-to-static]

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+  this->sources = { source };
+}
+
+QList<QJsonObject> GoogleTakeoutInputStream::readJson(const QString& source) {
```
warning: method 'readJson' can be made static [readability-convert-member-functions-to-static]

tsteven4 commented on googletakeout.cc:
```diff
@@ -0,0 +1,416 @@
+  return timeline;
+}
+
+QList<QString> GoogleTakeoutInputStream::readDir(const QString& source) {
```
warning: method 'readDir' can be made static [readability-convert-member-functions-to-static]

postmaxin committed:
Address code review requests

* add `googletakeout.h` to HEADERS
* convert most `#define`s to static members
* don't pollute global namespace
* rename logging functions
* convert some methods to static functions for clang-tidy
* `title_case` loop by reference instead of by index (clang-tidy)
* better use of debugging levels
* drop dead simplifiedRawPath code for now
* drop empty tracks
* default constructor for GoogleTakeoutInputStream
* add some tests that check the content of the data

Also, I found some unofficial documentation about the location history
format and added that (thanks @CarlosBergillos !)

postmaxin committed:
Merge branch 'master' into takeout2

postmaxin committed:
apply @tsteven4 's patch. thank you!!

postmaxin committed:
Merge branch 'takeout2' of github.com:postmaxin/gpsbabel into takeout2

postmaxin committed:
add license

postmaxin committed:
fix logging

postmaxin committed:
As I assigned the correct license to my code, I am skeptical that this makes a difference, but give Robert co-copyright.

postmaxin committed:
make the test more boring

postmaxin committed:
rm xml_slice

postmaxin committed:
revert constructor change from @tsteven4 's patch

postmaxin committed:
apply @tsteven4 's latest constructor

postmaxin committed:
Merge branch 'master' into takeout2

postmaxin committed:
Merge branch 'master' into takeout2

postmaxin committed:
apply @tsteven4 's latest patch -- `cmake --build . --target check` passes

postmaxin committed:
Merge branch 'takeout2' of github.com:postmaxin/gpsbabel into takeout2

tsteven4 committed:
Merge branch 'master' into takeout2
